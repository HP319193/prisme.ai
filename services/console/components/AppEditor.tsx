import {
  Button,
  Loading,
  notification,
  Schema,
  SchemaForm,
} from '@prisme.ai/design-system';
import { BlockLoader } from '@prisme.ai/blocks';
import api from '../utils/api';
import useAppConfig from '../utils/useAppConfig';
import { useWorkspace } from './WorkspaceProvider';
import { useWorkspaceLayout } from '../layouts/WorkspaceLayout/context';
import { useTranslation } from 'next-i18next';
import { useCallback } from 'react';

interface AppEditorProps {
  schema?: Schema;
  block?: string;
  appId: string;
}

const AppEditor = ({ schema, block, appId }: AppEditorProps) => {
  const {
    workspace: { id: workspaceId },
  } = useWorkspace();
  const { t } = useTranslation('workspaces');
  const { t: commmonT } = useTranslation('common');
  const { t: errorT } = useTranslation('errors');
  const { setDirty } = useWorkspaceLayout();
  const { appConfig, onAppConfigUpdate } = useAppConfig(workspaceId, appId);

  const onSubmit = async (value: any) => {
    try {
      await onAppConfigUpdate(value);
      notification.success({
        message: t('apps.saveSuccess'),
        placement: 'bottomRight',
      });
      setDirty(false);
    } catch (e) {
      notification.error({
        message: errorT('unknown', { errorName: e }),
        placement: 'bottomRight',
      });
    }
  };

  const onChange = useCallback(
    (value: any) => {
      const keys = Object.keys(value);
      const prevValue = keys.reduce(
        (prev, key) => ({
          ...prev,
          [key]: appConfig[key],
        }),
        {}
      );
      if (JSON.stringify(prevValue) === JSON.stringify(value)) return;
      setDirty(true);
    },
    [appConfig, setDirty]
  );

  if (!appConfig) return <Loading />;

  if (schema) {
    const s: Schema = {
      type: 'object',
      properties: schema as Schema['properties'],
    };

    return (
      <div className="p-6">
        <SchemaForm
          schema={s}
          onChange={onChange}
          onSubmit={onSubmit}
          initialValues={appConfig}
          buttons={[
            <div className="flex w-full justify-end mt-5" key="submit">
              <Button
                type="submit"
                className="!p-1"
                variant="primary"
                key="submit"
              >
                {commmonT('save')}
              </Button>
            </div>,
          ]}
        />
      </div>
    );
  }
  if (block) {
    return (
      <BlockLoader
        api={api}
        url={block}
        token={`${api.token}`}
        workspaceId={workspaceId}
        appInstance={appId}
        config={{}}
        appConfig={appConfig}
        onAppConfigUpdate={onAppConfigUpdate}
      />
    );
  }

  return null;
};

export default AppEditor;
