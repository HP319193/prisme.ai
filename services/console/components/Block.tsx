import * as React from 'react';
import * as prismeaiDS from '@prisme.ai/design-system';
import * as prismeaiSDK from '../utils/api';
import { ReactElement, useEffect, useState } from 'react';
import { nanoid } from 'nanoid';
import { useTranslation } from 'next-i18next';

if (process.browser) {
  // @ts-ignore
  window.__external = window.__external || {
    React: { ...React, default: React },
    prismeaiDS,
    prismeaiSDK,
  };
}

class BlockErrorBoundary extends React.Component {
  state = {
    hasError: false,
  };

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch(error: any, errorInfo: any) {
    // Vous pouvez aussi enregistrer l'erreur au sein d'un service de rapport.
    console.error(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      // Vous pouvez afficher n'importe quelle UI de repli.
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children;
  }
}

interface BlockComponentProps {
  entityId: string;
  token?: string;
  workspaceId?: string;
  appInstance?: string;
  language?: string;
}
type BlockComponent = (props: BlockComponentProps) => ReactElement;

interface BlockProps extends BlockComponentProps {
  url: string;
  renderLoading?: ReactElement;
}

export const ReactBlock = ({ url, renderLoading, ...props }: BlockProps) => {
  const {
    i18n: { language },
  } = useTranslation('workspaces');
  const [loading, setLoading] = useState(true);

  const [Component, setComponent] = useState<BlockComponent | null>(null);
  useEffect(() => {
    const uniqMethod = `__load_${nanoid()}`;
    // @ts-ignore
    window[uniqMethod] = (module) => {
      setComponent(() => {
        return module.default;
      });
      setLoading(false);
    };
    const s = document.createElement('script');

    s.innerText = `
    import * as module from '${url}';
    try {
      window['${uniqMethod}'](module);
    } catch (e) {}
    `;
    s.type = 'module';
    document.body.appendChild(s);

    return () => {
      // @ts-ignore
      delete window[uniqMethod];
    };
  }, [url]);

  return (
    <>
      {(loading && renderLoading) || null}
      {Component && <Component {...props} language={language} />}
    </>
  );
};

export const IFrameBlock = ({
  url,
  entityId,
  token,
  renderLoading,
}: BlockProps) => {
  const [loading, setLoading] = useState(true);
  const [height, setHeight] = useState(100);
  const handleLoad = React.useCallback(
    (e) => {
      setLoading(false);
      e.target.contentWindow.postMessage(
        {
          source: 'prisme.ai',
          token,
          id: entityId,
        },
        '*'
      );
    },
    [entityId, token]
  );
  useEffect(() => {
    const listener = (e: MessageEvent) => {
      if (e.data.source === entityId) {
        Object.keys(e.data).forEach((method) => {
          switch (method) {
            case 'init':
              setHeight(+e.data[method].height);
          }
        });
      }
    };
    window.addEventListener('message', listener);

    return () => {
      window.removeEventListener('message', listener);
    };
  }, [entityId, height]);

  return (
    <>
      {(loading && renderLoading) || null}
      <iframe
        src={url}
        onLoad={handleLoad}
        className="flex"
        style={{
          width: '100%',
          height: `${height}px`,
        }}
      />
    </>
  );
};

export const Block = ({ url, ...props }: BlockProps) => {
  const isJs = url.match(/\.js$/);
  if (isJs) {
    return (
      <BlockErrorBoundary>
        <ReactBlock url={url} {...props} />
      </BlockErrorBoundary>
    );
  }
  return <IFrameBlock url={url} {...props} />;
};

export default Block;
