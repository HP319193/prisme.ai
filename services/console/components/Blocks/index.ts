export { default as Title } from './Title';
export { default as Cards } from './Cards';
export { default as DataTable } from './DataTable';
