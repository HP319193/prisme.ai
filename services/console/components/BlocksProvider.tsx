import { FC, HTMLAttributes } from 'react';
import { Loading as DSLoading, SchemaForm } from '@prisme.ai/design-system';
import NextLink from 'next/link';
import externals from '../utils/externals';
import Image from 'next/image';
import { BlocksProvider as PRBlocksProvider } from '@prisme.ai/blocks';
import down from '../icons/down.svg';
import { usePageEndpoint } from '../utils/urls';

const Loading = () => (
  <DSLoading className="bg-white absolute top-0 right-0 bottom-0 left-0" />
);
const Link = ({
  href = '',
  ...props
}: { href: string } & HTMLAttributes<HTMLAnchorElement>) => {
  return <NextLink {...props} href={href || ''} />;
};
const DownIcon = ({ className }: { className?: string }) => (
  <Image src={down.src} width={14} height={14} alt="" className={className} />
);

export const BlocksProvider: FC = ({ children }) => {
  const pageHost = usePageEndpoint();
  return (
    <PRBlocksProvider
      externals={externals}
      components={{ Link, Loading, DownIcon, SchemaForm }}
      utils={{
        getWorkspaceHost() {
          return pageHost;
        },
      }}
    >
      {children}
    </PRBlocksProvider>
  );
};

export const PublicBlocksProvider: FC = ({ children }) => {
  return (
    <PRBlocksProvider
      externals={externals}
      components={{ Link, Loading, DownIcon, SchemaForm }}
    >
      {children}
    </PRBlocksProvider>
  );
};
