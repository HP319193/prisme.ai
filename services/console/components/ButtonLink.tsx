import Link from "next/link"
import { ButtonHTMLAttributes, FC } from "react"

interface ButtonLinkProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  href?: string;
}
export const ButtonLink: FC<ButtonLinkProps> = ({ children, href, ...props }) => {

  return <button style={{ background: 'none', border: 0, cursor: 'pointer', fontSize: 'inherit', ...props.style }} {...props}>
    {href && <Link href={href}><>{children}</></Link>}
    {!href && children}
  </button>
}

export default ButtonLink
