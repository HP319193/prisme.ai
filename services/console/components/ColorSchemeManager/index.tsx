import Storage from '../../utils/Storage';

export function getColorScheme() {
  return (
    (document.documentElement.dataset.colorScheme as 'dark' | 'light') ||
    Storage.get('color-scheme') ||
    (window.matchMedia('(prefers-color-scheme: dark)').matches
      ? 'dark'
      : 'light')
  );
}

export function setColorScheme(scheme: 'dark' | 'light', save: boolean = true) {
  if (save) {
    Storage.set('color-scheme', scheme);
  }
  document.documentElement.setAttribute('data-color-scheme', scheme);
}

export function toggleColorScheme() {
  const next = getColorScheme() === 'dark' ? 'light' : 'dark';
  setColorScheme(next);
}

export const ColorSchemeManager = () => {
  if (typeof window === 'undefined') return null;

  setColorScheme(getColorScheme(), false);

  // @ts-ignore
  window.Prisme = window.Prisme || {};
  // @ts-ignore
  window.Prisme.ai = window.Prisme.ai || {};
  // @ts-ignore
  window.Prisme.ai.colorScheme = {
    get: getColorScheme,
    set: setColorScheme,
    toggle: toggleColorScheme,
  };
  return null;
};

export default ColorSchemeManager;
