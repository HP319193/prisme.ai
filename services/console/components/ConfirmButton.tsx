import { Button, Tooltip } from '@prisme.ai/design-system';
import { ButtonProps } from '@prisme.ai/design-system/lib/Components/Button';
import { TooltipProps } from 'antd';
import { useTranslation } from 'next-i18next';
import { ReactNode, useRef } from 'react';

interface ConfirmButtonProps extends ButtonProps {
  children: ReactNode;
  onConfirm: () => void;
  confirmLabel?: string;
  yesLabel?: string;
  noLabel?: string;
}
export const ConfirmButton = ({
  children,
  onConfirm,
  confirmLabel,
  yesLabel,
  noLabel,
  ...props
}: ConfirmButtonProps) => {
  const tooltipRef = useRef<{ close: () => void }>(null);
  const { t } = useTranslation('common');
  return (
    <Tooltip
      ref={tooltipRef}
      title={
        <div>
          <div>{confirmLabel}</div>
          <div className="flex justify-end">
            <Button
              type="button"
              variant="grey"
              onClick={() => {
                onConfirm();
                tooltipRef.current?.close();
              }}
            >
              {yesLabel || t('yes')}
            </Button>
            <Button type="button" onClick={() => tooltipRef.current?.close()}>
              {noLabel || t('no')}
            </Button>
          </div>
        </div>
      }
      trigger={['click']}
      color="var(--warning-color)"
    >
      <Button className="!text-warning" {...props}>
        {children}
      </Button>
    </Tooltip>
  );
};
export default ConfirmButton;
