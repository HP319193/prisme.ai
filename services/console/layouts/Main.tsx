import Link from 'next/link';
import { Button } from 'primereact/button';
import { FC, ReactNode } from 'react';
import Loading from '../components/Loading';
import { useUser } from '../components/UserProvider';
import Toaster from './Toaster';

interface MainProps {
  title?: string;
  leftContent?: ReactNode;
  rightContent?: ReactNode;
}
export const Main: FC<MainProps> = ({
  children,
  title,
  leftContent,
  rightContent,
}) => {
  const { user, loading, signout } = useUser();

  if (!user && loading) {
    return (
      <div className="flex min-h-screen max-h-screen flex-column justify-center">
        <Loading className="text-5xl" />
      </div>
    );
  }

  return (
    <div className="flex min-h-screen max-h-screen flex-column">
      <div className="bg-primary p-4 z-1 drop-shadow-md">
        <div className="flex justify-between">
          <div className="flex direction-row align-center">
            <div className="mr-2">
              <Link href="/workspaces">{title || 'Prisme.ai'}</Link>
            </div>
            {leftContent}
          </div>
          <div>
            {rightContent && rightContent}
            <Button icon="pi pi-power-off" onClick={signout} />
          </div>
        </div>
      </div>
      <div
        className="flex flex-1 surface-ground relative"
        style={{ overflow: 'auto', overflowX: 'hidden' }}
      >
        {children}
      </div>
    </div>
  );
};

export default Main;
