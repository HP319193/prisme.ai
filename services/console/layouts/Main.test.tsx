import Main from "./Main";
import renderer from "react-test-renderer";
import { useUser } from "../components/UserProvider";

jest.mock("../components/UserProvider", () => {
  const mock = { user: undefined, loading: true, signout: jest.fn() };
  return {
    useUser: () => mock,
  };
});

it("should render loading", () => {
  const root = renderer.create(<Main />);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should render", () => {
  useUser().user = {
    firstName: "",
  };
  useUser().loading = false;
  const root = renderer.create(<Main />);
  expect(root.toJSON()).toMatchSnapshot();
});
