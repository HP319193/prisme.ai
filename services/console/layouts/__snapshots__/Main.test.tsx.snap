// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="flex min-h-screen max-h-screen flex-col"
>
  <div
    className="bg-primary p-4 z-1 "
  >
    <div
      className="flex justify-between"
    >
      <div
        className="flex direction-row align-center"
      >
        <div
          className="mr-2"
        >
          <a
            href="/workspaces"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Prisme.ai
          </a>
        </div>
      </div>
      <div>
        <button
          className="p-button p-component p-button-icon-only"
          disabled={false}
          onClick={[MockFunction]}
        >
          <span
            className="p-button-icon p-c pi pi-power-off"
          />
          <span
            className="p-button-label p-c"
            dangerouslySetInnerHTML={
              Object {
                "__html": "&nbsp;",
              }
            }
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="flex flex-1 surface-ground relative"
    style={
      Object {
        "overflow": "auto",
        "overflowX": "hidden",
      }
    }
  />
</div>
`;

exports[`should render loading 1`] = `
<div
  className="flex min-h-screen max-h-screen flex-col justify-center"
>
  <div
    className="flex flex-1 align-center justify-center"
  >
    <div
      className="text-5xl align-self-center justify-self-center pi pi-spin pi-spinner"
    />
  </div>
</div>
`;
